### PR TITLE
Sandra Merges maps and makes changes to events

### PR DESCRIPTION
**Pull request by ServingCoder**
Sandra adds maps functionality. 
Sandra makes changes to events, adds checkedInAttendeeLocations which holds a list of Location objects

### CODE REVIEW

File: code/noRAM/app/build.gradle.kts [hunk 1]
```diff
@@ -87,4 +87,13 @@ dependencies {
     // Confetti Library
     implementation("nl.dionsegijn:konfetti-xml:2.0.4")
 
+    //Maps API
+    implementation("org.osmdroid:osmdroid-android:6.1.18")
+    //Location Google Play API
+    implementation("com.google.android.gms:play-services-location:21.2.0")
+    //Map osmdroid-geopackage support, I am not ever sure I need this
+    //I am not sure the version number is correct
```
Comment (WhoShotCupidd) [inline code/noRAM/app/build.gradle.kts]: can we remove these?

File: code/noRAM/app/src/main/java/com/example/noram/AttendeeEventInfoActivity.java [hunk 2]
```diff
@@ -52,6 +52,13 @@ public class AttendeeEventInfoActivity extends AppCompatActivity {
     private void signup(){
         // TODO: update database to add signed-in attendees to event
         // TODO: send to message page: should send to signed-in page instead of checked-in page
+        // sign-in the event and display sign-in message
+        EventManager.checkInToEvent(event.getId(), null);
+        Toast.makeText(this, "Successfully checked in!", Toast.LENGTH_SHORT).show();
```
Comment (WhoShotCupidd) [inline code/noRAM/app/src/main/java/com/example/noram/AttendeeEventInfoActivity.java:65]: toast message?

File: code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java [hunk 3]
```diff
@@ -1,23 +1,159 @@
 package com.example.noram;
```
Comment (WhoShotCupidd) [inline code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java:8]: file header

File: code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java [hunk 4]
```diff
@@ -1,23 +1,159 @@
...
-
     /**
      * Setup the activity when it is created.
      * @param savedInstanceState If the activity is being re-initialized after
      *    previously being shut down then this Bundle contains the data
      */
+    private final int REQUEST_PERMISSIONS_REQUEST_CODE = 1;
+    private MapView map = null;
+
     @Override
```
Comment (WhoShotCupidd) [inline code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java:89]: javadoc

File: code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java [hunk 5]
```diff
@@ -1,23 +1,159 @@
...
+        }
+        else{
+            //Tell user problem
+            Toast.makeText(getApplicationContext(), "Must allow Location to use maps.", Toast.LENGTH_SHORT).show();
+            //exit maps activity somehow
+            finish();
+            }
+
+    }
+    @SuppressLint("MissingPermission")
```
Comment (WhoShotCupidd) [inline code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java:151]: java doc

File: code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java [hunk 6]
```diff
@@ -1,23 +1,159 @@
...
+    private void getLastLocationUser() {
+        FusedLocationProviderClient fusedLocationClient = LocationServices.getFusedLocationProviderClient(this);
+        fusedLocationClient.getLastLocation().addOnSuccessListener(this, new OnSuccessListener<Location>() {
+            @Override
+            public void onSuccess(Location location) {
```
Comment (WhoShotCupidd) [inline code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java:163]: java doc

File: code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java [hunk 7]
```diff
@@ -1,23 +1,159 @@
...
+                    //only the organizer gets a special call to handleLocation with drawable id
+                    handleLocation(location, R.drawable.attendee_map_icon);
+                } else {
+                    // Location is null, handle the case accordingly
+                    Toast.makeText(getApplicationContext(), "Unable to retrieve location", Toast.LENGTH_SHORT).show();
+                }
+            }
+        }).addOnFailureListener(this, new OnFailureListener() {
+            @Override
+            public void onFailure(@NonNull Exception e) {
```
Comment (WhoShotCupidd) [inline code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java:180]: javadoc

File: code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java [hunk 8]
```diff
@@ -1,23 +1,159 @@
...
+                // Failed to get location
+                e.printStackTrace();
+                Toast.makeText(getApplicationContext(), "Failed to get location", Toast.LENGTH_SHORT).show();
+            }
+        });
+    }
+    private void handleLocation(Location location, int drawID) {
```
Comment (WhoShotCupidd) [inline code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java]: javadoc

File: code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java [hunk 9]
```diff
@@ -1,23 +1,159 @@
...
+
+        GeoPoint point = new GeoPoint(latitude, longitude);
+        Marker startMarker = new Marker(map);
+        startMarker.setPosition(point);
+        startMarker.setAnchor(Marker.ANCHOR_CENTER, Marker.ANCHOR_CENTER);
+        startMarker.setIcon(getResources().getDrawable(drawID));
+        map.getOverlays().add(startMarker);
+        map.getController().setCenter(point);
+    }
+    private void handleLocation(Location location) {
```
Comment (WhoShotCupidd) [inline code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java]: javadoc

File: code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java [hunk 10]
```diff
@@ -1,23 +1,159 @@
...
+        Double longitude = location.getLongitude();
+
+        GeoPoint point = new GeoPoint(latitude, longitude);
+        Marker startMarker = new Marker(map);
+        startMarker.setPosition(point);
+        startMarker.setAnchor(Marker.ANCHOR_CENTER, Marker.ANCHOR_CENTER);
+        map.getOverlays().add(startMarker);
+        map.getController().setCenter(point);
+    }
+    private boolean hasLocationPermissions() {
```
Comment (WhoShotCupidd) [inline code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java:227]: javadoc

File: code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java [hunk 11]
```diff
@@ -1,23 +1,159 @@
...
+        return (checkSelfPermission(Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED && checkSelfPermission(Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED);
+    }
+
+    @Override
```
Comment (WhoShotCupidd) [inline code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java:254]: javadoc

File: code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java [hunk 12]
```diff
@@ -1,23 +1,159 @@
...
+    public void onResume() {
+        super.onResume();
+            if (map != null) {
+                map.onResume();
+            }
+    }
+    @Override
```
Comment (WhoShotCupidd) [inline code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java:271]: javadoc

File: code/noRAM/app/src/main/java/com/example/noram/controller/EventManager.java [hunk 13]
```diff
@@ -2,14 +2,24 @@
```
Comment (WhoShotCupidd) [inline code/noRAM/app/src/main/java/com/example/noram/controller/EventManager.java:6]: not sure if this is your file but it needs a header.

Comment (WhoShotCupidd) [reply code/noRAM/app/src/main/java/com/example/noram/controller/EventManager.java:6]: @gdumouli u need to add this

**Review by WhoShotCupidd — CHANGES_REQUESTED** [review]
please look at my comments

File: code/noRAM/app/src/main/java/com/example/noram/QrScanFragment.java [hunk 14]
```diff
@@ -140,7 +145,14 @@ private void processQRCode(String qrCodeString) {
                 QRType qrType = QRType.valueOf(qrDocument.getString("type"));
 
                 if (qrType == QRType.SIGN_IN) {
-                    EventManager.checkInToEvent(eventId);
+                    Location attendeeLocation = null;
+                    if (MainActivity.attendee.getAllowLocation()) {
+                        FusedLocationProviderClient fusedLocationClient = LocationServices.getFusedLocationProviderClient(getActivity());
+                        //TODO add a onsuccess and onfail listener
+                        //grabbing current loaction of the phone
+                        attendeeLocation = fusedLocationClient.getLastLocation().getResult();
```
Comment (ColeDewis) [inline code/noRAM/app/src/main/java/com/example/noram/QrScanFragment.java]: this line does not work - "Task is not yet completed" - you probably need a callback

**Review by ColeDewis — DISMISSED** [review]
Haven't looked throuhg the code in detail since christiaan did, but there's a crash. 

Also, there's no back button/top bar on the map page, though the map loads and looks good

File: code/noRAM/app/src/main/java/com/example/noram/QrScanFragment.java (hunk 14 above)
Comment (WhoShotCupidd) [reply code/noRAM/app/src/main/java/com/example/noram/QrScanFragment.java]: thx for running it cole :)

File: code/noRAM/app/build.gradle.kts (hunk 1 above)
Comment (ServingCoder) [reply code/noRAM/app/build.gradle.kts]: I removed them. If we need them I will add them back later

File: code/noRAM/app/src/main/java/com/example/noram/AttendeeEventInfoActivity.java (hunk 2 above)
Comment (ServingCoder) [reply code/noRAM/app/src/main/java/com/example/noram/AttendeeEventInfoActivity.java:65]: @Semicarlin I think this is his

Comment (ServingCoder) [reply code/noRAM/app/src/main/java/com/example/noram/AttendeeEventInfoActivity.java:65]: What I added was passing 'null' to  EventManager.checkInToEvent(event.getId(), null)

File: code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java (hunk 3 above)
Comment (ServingCoder) [reply code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java:8]: done

File: code/noRAM/app/src/main/java/com/example/noram/AttendeeEventInfoActivity.java (hunk 2 above)
Comment (WhoShotCupidd) [reply code/noRAM/app/src/main/java/com/example/noram/AttendeeEventInfoActivity.java:65]: @ServingCoder can u push that work plz so we can see it

Comment (ServingCoder) [reply code/noRAM/app/src/main/java/com/example/noram/AttendeeEventInfoActivity.java:65]: @WhoShotCupidd I am making changes on my computer and will make another commit soon. Should I make my changes in a different manner?

Comment (WhoShotCupidd) [reply code/noRAM/app/src/main/java/com/example/noram/AttendeeEventInfoActivity.java:65]: ohh nah thats fine just normally if you change something to stop a bug you push it so the reviewer can see that code and see if that actually resolves the issue. Really doesn't matter tho :) @ServingCoder

Comment (ServingCoder) [reply code/noRAM/app/src/main/java/com/example/noram/AttendeeEventInfoActivity.java:65]: Ok, I'll do two pushes. One with all the comment stuff. Another one that addresses the bug. @WhoShotCupidd

File: code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java [hunk 15]
```diff
@@ -1,23 +1,220 @@
...
+import org.osmdroid.views.overlay.mylocation.GpsMyLocationProvider;
+import org.osmdroid.views.overlay.mylocation.MyLocationNewOverlay;
+
+import java.util.ArrayList;
+import java.util.Objects;
 
 /**
  * The OrganizerEventMapActivity class displays the map of an event for the organizer and check in locations.
  * A {@link AppCompatActivity} subclass.
+ * @author Sandra Taskovic
```
Comment (WhoShotCupidd) [inline code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java:67]: add maintainer (you)

File: code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java [hunk 16]
```diff
@@ -1,23 +1,220 @@
...
+ *         https://github.com/osmdroid
+ *         CHATGPT for adding onSuccesListener and onFailureListner when acessing user location
+ *         this video for help on how to use googlePlay API https://www.youtube.com/watch?v=M0kUd2dpxo4&t=854s
  */
 public class OrganizerEventMapActivity extends AppCompatActivity {
+    /**
+     * Setup the map activity.
+     * A {@link AppCompatActivity} subclass.
```
Comment (WhoShotCupidd) [inline code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java:66]: why is this here?

File: code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java (hunk 6 above)
Comment (WhoShotCupidd) [reply code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java:163]: you never added a javadoc?

File: code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java [hunk 17]
```diff
@@ -1,23 +1,220 @@
...
+    /**
+     * Once the location is acessed, draw it on the map
+     * @param location the location of the user
+     * @param drawID the drawable ID location used to draw location on the map
+     */
+    private void handleLocation(Location location, int drawID) {
+        Double latitude = location.getLatitude();
+        Double longitude = location.getLongitude();
+
+        GeoPoint point = new GeoPoint(latitude, longitude);
```
Comment (WhoShotCupidd) [inline code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java:165]: can we get just a few more comments here so we can follow pwetty pwez

File: code/noRAM/app/src/main/java/com/example/noram/SandraTest.java
Comment (WhoShotCupidd) [inline code/noRAM/app/src/main/java/com/example/noram/SandraTest.java:1]: is there a reason for this file?

File: code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java (hunk 15 above)
Comment (ServingCoder) [reply code/noRAM/app/src/main/java/com/example/noram/OrganizerEventMapActivity.java:67]: sure

File: code/noRAM/app/src/main/java/com/example/noram/QrScanFragment.java [hunk 18]
```diff
@@ -149,7 +162,39 @@ private void processQRCode(String qrCodeString) {
...
+                } else {
+                    //TODO: exit maps. User provided permission to use location but the location was not properly received.
+                    //TODO: create an internal error message
+                    Toast.makeText(getContext(), "Unable to retrieve location, signing in without location", Toast.LENGTH_SHORT).show();
+                    EventManager.checkInToEvent(ID, null);
+                }
+            }
+        }).addOnFailureListener(requireActivity(), new OnFailureListener() {
+            @Override
+            public void onFailure(@NonNull Exception e) {
```
Comment (WhoShotCupidd) [inline code/noRAM/app/src/main/java/com/example/noram/QrScanFragment.java:209]: javadoc

File: code/noRAM/app/src/main/java/com/example/noram/QrScanFragment.java [hunk 19]
```diff
@@ -149,7 +162,39 @@ private void processQRCode(String qrCodeString) {
...
+     */
+    @SuppressLint("MissingPermission")
+    private void getLocationQREntry(String ID) {
+        //grab the FusedLocationProviderClient builder
+        FusedLocationProviderClient fusedLocationClient = LocationServices.getFusedLocationProviderClient(requireContext());
+        //attempt to get the location
+        fusedLocationClient.getLastLocation().addOnSuccessListener(requireActivity(), new OnSuccessListener<Location>() {
+            // if the API can get a location
+            @Override
+            public void onSuccess(Location location) {
```
Comment (WhoShotCupidd) [inline code/noRAM/app/src/main/java/com/example/noram/QrScanFragment.java:189]: javadoc

**Review by WhoShotCupidd — CHANGES_REQUESTED** [review]
add all the java doc I have listed here then its good

**Comment by ServingCoder** [comment]
@ColeDewis Please let me know if you'd like me to move the location permission prompt outside the manifest and into the map?

**Comment by ServingCoder** [comment]
@ColeDewis  @WhoShotCupidd  I will fix the EventValidatorUnitTest because the event object has more attributes now.

File: code/noRAM/app/src/androidTest/java/com/example/noram/MainActivityTest.java [hunk 20]
```diff
@@ -52,6 +52,25 @@ public void setup() throws InterruptedException {
     /**
      * Verifies the attendee "browse events" button opens the attendee page
      */
+//TODO: do we need this still? I did not write this test
+
```
Comment (WhoShotCupidd) [inline code/noRAM/app/src/androidTest/java/com/example/noram/MainActivityTest.java]: this needs to be here

**Review by WhoShotCupidd — APPROVED** [review]
lgtm
